server/client.go

```diff
@@ -810,8 +811,18 @@ func (c *client) flushOutbound() bool {
 	attempted := c.out.pb
 	apm := c.out.pm
 
-	// Do NOT hold lock during actual IO
-	c.mu.Unlock()
+	// What we are doing here is seeing if we are getting behind. This is
+	// generally not a gradual thing and will spike quickly. Use some basic
+	// logic to try to understand when this is happening through no fault of
+	// our own. How we attempt to get back into a more balanced state under
+	// load will be to hold our lock during IO, forcing others to wait and
+	// applying back pressure to the publishers sending to us.
+	releaseLock := c.out.pb < maxBufSize*4
+
+	// Do NOT hold lock during actual IO unless we are behind
+	if releaseLock {
+		c.mu.Unlock()
+	}
```

Comment:
Can't reproduce the slow consumer issue with this change now (tried with `nats-bench -np 40 -ns 1 -n 100000000 -ms 1024 foo`)

Reply:
Yes I was using the new fan in tests but glad you tried this one too. Thanks!